styx-common/src/main/java/com/spotify/styx/util/CachedSupplier.java

```diff
@@ -20,91 +20,41 @@
  */
 public class CachedSupplier<T> implements Supplier<T> {
 
-  private static final Logger LOG = LoggerFactory.getLogger(CachedSupplier.class);
-  private static final long DEFAULT_TIMEOUT_MILLIS = 30_000;
+  private static final Duration DEFAULT_TIMEOUT = Duration.ofSeconds(30);
 
-  private final ThrowingSupplier<T, Exception> delegate;
-  private final Time time;
-  private final long timeoutMillis;
+  private final LoadingCache<Boolean, T> cache;
 
-  private final AtomicReference<T> cachedValue = new AtomicReference<>();
-  private final Object lock = new Object() {};
-  private volatile long cacheTime;
-
-  public CachedSupplier(ThrowingSupplier<T, Exception> delegate, Time time) {
-    this(delegate, time, DEFAULT_TIMEOUT_MILLIS);
+  public CachedSupplier(Try.CheckedSupplier<T> delegate, Time time) {
+    this(delegate, time, DEFAULT_TIMEOUT);
   }
 
-  public CachedSupplier(ThrowingSupplier<T, Exception> delegate, Time time, long timeoutMillis) {
-    this.delegate = Objects.requireNonNull(delegate);
-    this.time = Objects.requireNonNull(time);
-    this.timeoutMillis = timeoutMillis;
-
-    // TODO: use System.nanoTime instead.
-    cacheTime = time.get().toEpochMilli();
+  CachedSupplier(Try.CheckedSupplier<T> delegate, Time time, Duration timeout) {
+    this.cache = CacheBuilder.newBuilder()
+        .refreshAfterWrite(timeout)
+        .ticker(new Ticker() {
+          @Override
+          public long read() {
+            return time.nanoTime();
+          }
+        })
+        .build(CacheLoader.from(() -> Try.of(delegate).get()));
   }
 
   @Override
   public T get() {
-    T value = cachedValue.get();
-
-    if (value != null && !timedOut()) {
-      return value;
-    }
-
-    return getSynchronized();
-  }
-
-  /**
-   * Synchronize on updating the cached value to avoid potentially many redundant and concurrent fetches/computations.
-   * The use of {@link CachedSupplier} is a strong indicator that fetching/computing the value is expensive and hence
-   * it makes sense to try hard to avoid doing it too much.
-   */
-  private T getSynchronized() {
-    synchronized (lock) {
-      final T value = cachedValue.get();
-      if (value != null && !timedOut()) {
-        return value;
-      }
-      try {
-        final T newValue = delegate.get();
-        cachedValue.set(newValue);
-        cacheTime = time.get().toEpochMilli();
-        return newValue;
-      } catch (Throwable e) {
-        if (value == null) {
-          throw Throwables.propagate(e);
-        } else {
-          LOG.warn("Failed to update from delegate supplier, using old value", e);
-          return value;
-        }
-      }
-    }
-  }
-
-  private boolean timedOut() {
-    return time.get().toEpochMilli() - cacheTime > timeoutMillis;
-  }
-
-  @FunctionalInterface
-  public interface ThrowingSupplier<T, E extends Exception> {
-    T get() throws E;
+    return Try.of(() -> cache.get(Boolean.TRUE)).get();
```

Comment:
This changes existing behaviour I think. Somewhere might get unchecked exception.

Reply:
Any exception from `get()` was already unchecked though.

Reply:
I mean if it failed to refresh the value it fell back to old value if not null.

Reply:
Ah. Yeah now this PR also returns the old value if refresh fails.